core/chains/evm/logpoller/log_poller.go

```diff
@@ -0,0 +1,430 @@
+		// We truncate all the blocks and logs after the LCA.
+		// We could preserve the logs for forensics, since its possible
+		// that applications see them and take action upon it, however that
+		// results in significantly slower reads since we must then compute
+		// the canonical set per read. Typically if an application took action on a log
+		// it would be saved elsewhere e.g. eth_txes, so it seems better to just support the fast reads.
+		// Its also nicely analogous to reading from the chain itself.
+		err3 = lp.orm.q.Transaction(func(tx pg.Queryer) error {
+			// These deletes are bounded by reorg depth, so they are
+			// fast and should not slow down the log readers.
+			err3 = lp.orm.DeleteRangeBlocks(lca+1, latest, pg.WithQueryer(tx))
+			if err3 != nil {
+				lp.lggr.Warnw("Unable to clear reorged blocks, retrying", "err", err3)
+				return err3
+			}
+			err3 = lp.orm.DeleteLogs(lca+1, latest, pg.WithQueryer(tx))
+			if err3 != nil {
+				lp.lggr.Warnw("Unable to clear reorged logs, retrying", "err", err3)
+				return err3
+			}
+			return nil
+		})
+		if err3 != nil {
+			// If we crash or fail to update state we simply do not increment current so we'll detect the same
+			// reorg (if still present) and retry.
+			return false, 0, err3
+		}
+		return true, lca + 1, nil
+	}
+	return false, 0, nil
+}
+
+// PollAndSaveLogs On startup/crash current is the first block after the last processed block.
+func (lp *LogPoller) PollAndSaveLogs(ctx context.Context, current int64) int64 {
+	lp.lggr.Infow("Polling for logs", "current", current)
+	// Get latest block on chain
+	latestBlock, err1 := lp.ec.BlockByNumber(ctx, nil)
+	if err1 != nil {
+		lp.lggr.Warnw("Unable to get latest block", "err", err1, "current", current)
+		return current
+	}
+	latest := latestBlock.Number().Int64()
+	if current > latest {
+		lp.lggr.Debugw("No new blocks since last poll", "current", current, "latest", current)
+		return current
+	}
+	// Get current block
+	block, err1 := lp.ec.BlockByNumber(ctx, big.NewInt(current))
+	if err1 != nil {
+		lp.lggr.Warnw("Unable to get block", "err", err1, "current", current)
+		return current
+	}
+	// Possibly handle a reorg
+	reorgDetected, newPollBlock, err1 := lp.maybeHandleReorg(block, current, latest)
+	if err1 != nil {
+		// Continuously retry from same block on any error in reorg handling.
+		return current
+	}
+	// If we did detect a reorg, we'll have a new block number to start from (LCA+1)
+	// so let's resume polling from there.
+	if reorgDetected {
+		current = newPollBlock
+	}
+
+	// Backfill finalized blocks if we can for performance.
+	// E.g. 1<-2<-3(current)<-4<-5<-6<-7(latest), finality is 2. So 3,4,5 can be batched.
+	// start = current = 3, end = latest - finality = 7-2 = 5 (inclusive range).
+	if (latest - current) >= lp.finalityDepth {
+		lp.lggr.Infow("Backfilling logs", "start", current, "end", latest-lp.finalityDepth)
+		current = lp.backfill(ctx, current, latest-lp.finalityDepth)
+	}
```

Comment:
nit: this code appears to be a repetition of the for loop body, except for the backfill part

Reply:
not quite - we need to take different action on the reorgDetected in the loop. I can combine maybeHandleReorg and the BlockByNumber at the expense of a little clarity though